apollo-router/src/plugins/diagnostics/system_info/mod.rs

```diff
@@ -0,0 +1,631 @@
+        "arm" => "arm32",
+        "powerpc" => "ppc32",
+        "powerpc64" => "ppc64",
+        arch => arch,
+    }
+}
+
+/// Detect effective CPU count considering cgroup limits (improved from fleet detector)
+#[cfg(not(target_os = "linux"))]
+fn detect_effective_cpu_count(system_cpus: u64) -> (&'static str, u64) {
+    ("system", system_cpus)
+}
+
+#[cfg(target_os = "linux")]
+fn detect_effective_cpu_count(system_cpus: u64) -> (&'static str, u64) {
+    // Determine cgroup version first
+    match std::fs::read_to_string("/proc/filesystems").map(|fs| detect_cgroup_version(&fs)) {
+        Ok(CGroupVersion::CGroup2) => {
+            // cgroup v2: read from cpu.max
+            match std::fs::read_to_string("/sys/fs/cgroup/cpu.max") {
+                Ok(contents) => {
+                    if contents.starts_with("max") {
+                        ("system", system_cpus)
+                    } else {
+                        match contents.split_once(' ') {
+                            Some((quota, period)) => (
+                                "cgroup2",
+                                calculate_cpu_count_with_default(system_cpus, quota, period),
+                            ),
+                            None => ("system", system_cpus),
+                        }
+                    }
+                }
+                Err(_) => ("system", system_cpus),
+            }
+        }
+        Ok(CGroupVersion::CGroup) => {
+            // cgroup v1: read from separate quota and period files
+            let quota = std::fs::read_to_string("/sys/fs/cgroup/cpu/cpu.cfs_quota_us")
+                .map(|s| s.trim().to_string())
+                .ok();
+            let period = std::fs::read_to_string("/sys/fs/cgroup/cpu/cpu.cfs_period_us")
+                .map(|s| s.trim().to_string())
+                .ok();
+
+            match (quota, period) {
+                (Some(quota), Some(period)) => {
+                    if quota == "-1" {
+                        ("system", system_cpus)
+                    } else {
+                        (
+                            "cgroup",
+                            calculate_cpu_count_with_default(system_cpus, &quota, &period),
+                        )
+                    }
+                }
+                _ => ("system", system_cpus),
+            }
+        }
+        _ => ("system", system_cpus),
+    }
+}
+
+/// Detect cgroup version from /proc/filesystems content
+#[cfg(target_os = "linux")]
+fn detect_cgroup_version(filesystems: &str) -> CGroupVersion {
+    use std::collections::HashSet;
+
+    let versions: HashSet<_> = filesystems
+        .lines()
+        .flat_map(|line| line.split_whitespace())
+        .filter(|word| word.contains("cgroup"))
+        .collect();
+
+    if versions.contains("cgroup2") {
+        CGroupVersion::CGroup2
+    } else if versions.contains("cgroup") {
+        CGroupVersion::CGroup
+    } else {
+        CGroupVersion::None
+    }
```

Comment:
This can just be a `.find_map`.

Reply:
`find_map` will find the first match but cgroup2 may appear after cgroup.

Reply:
Ah, good point. I suppose this could be done with mapping to `CGroupVersion` and then `max` (you just need to derive `Ord`). Though, that has other downsides...